### PR TITLE
Fix missing warningstream (or similar problem)

### PR DESCRIPTION
**Pull request by HybridDog**
This is a fix for #6861.
Instead of a minetest setting, a command line parameter is used to en-/disable coloured logs.

### CODE REVIEW

File: src/log.h [hunk 1]
```diff
@@ -119,9 +121,8 @@ class StreamLogOutput : public ICombinedLogOutput {
 
 	void logRaw(LogLevel lev, const std::string &line)
 	{
-		static const std::string use_logcolor = g_settings->get("log_color");
-
-		bool colored = use_logcolor == "detect" ? is_tty : use_logcolor == "yes";
+		bool colored = Logger::use_logcolor == "auto" ? is_tty :
+			Logger::use_logcolor == "always";
```
Comment (rubenwardy) [inline src/log.h]: I'd put explicit parentheses around the equivalence expressions just to make the order of operations explicit

Comment (HybridDog) [reply src/log.h]: ok

File: src/log.h [hunk 2]
```diff
@@ -119,9 +121,8 @@ class StreamLogOutput : public ICombinedLogOutput {
 
 	void logRaw(LogLevel lev, const std::string &line)
 	{
-		static const std::string use_logcolor = g_settings->get("log_color");
-
-		bool colored = use_logcolor == "detect" ? is_tty : use_logcolor == "yes";
+		bool colored = (Logger::use_logcolor == "auto") ? is_tty :
```
Comment (nerzhul) [inline src/log.h]: ouch, and ugly condition on each log. Don't use a string but a enum on u8 instead

Comment (HybridDog) [reply src/log.h]: A simple boolean should be sufficient, is_tty is set only once for each stream after all.

File: src/main.cpp [hunk 3]
```diff
@@ -257,6 +257,9 @@ static void set_allowed_options(OptionList *allowed_options)
 			_("Set world by name (implies local game)"))));
 	allowed_options->insert(std::make_pair("quiet", ValueSpec(VALUETYPE_FLAG,
 			_("Print to console errors only"))));
+	allowed_options->insert(std::make_pair("color", ValueSpec(VALUETYPE_STRING,
```
Comment (red-001) [inline src/main.cpp]: maybe the preprocessor could be used to not include the `auto` option in windows builds if it doesn't work?

File: src/main.cpp [hunk 4]
```diff
@@ -382,6 +385,13 @@ static void setup_log_params(const Settings &cmd_args)
 		g_logger.addOutputMaxLevel(&stderr_output, LL_ERROR);
 	}
 
+	// Coloured log messages (see log.h)
+	if (cmd_args.exists("color")) {
+		std::string mode = cmd_args.get("color");
+		Logger::color_mode = (mode == "auto") ? LOG_COLOR_AUTO :
+			(mode == "always") ? LOG_COLOR_ALWAYS : LOG_COLOR_NEVER;
```
Comment (red-001) [inline src/main.cpp]: This looks messy imho.

File: src/main.cpp (hunk 3 above)
Comment (HybridDog) [reply src/main.cpp]: Nice idea. I didn't exclude it everywhere because at some point in time

File: src/main.cpp (hunk 4 above)
Comment (HybridDog) [reply src/main.cpp]: l agree. I put if statements instead.

File: src/log.h [hunk 5]
```diff
@@ -111,18 +118,17 @@ class StreamLogOutput : public ICombinedLogOutput {
 		m_stream(stream)
 	{
 #if !defined(_WIN32)
-		is_tty = isatty(fileno(stdout));
+		colored = (Logger::color_mode == LOG_COLOR_ALWAYS) ||
+			(Logger::color_mode == LOG_COLOR_AUTO && isatty(fileno(stdout)));
 #else
-		is_tty = false;
+		colored = Logger::color_mode == COLOR_ALWAYS;
```
Comment (SmallJoker) [inline src/log.h]: `== LOG_COLOR_ALWAYS`

Comment (HybridDog) [reply src/log.h]: fixed

**Comment by lhofhansl** [comment]
Looks like this does not work. I usually have a white terminal and there is currently no way to turn colored logs off (passing --color never does nothing). This is actually quite vexing. Yellow logs on white background are 100% unreadable.

**Comment by HybridDog** [comment]
It doesn't work for me too. Maybe I forgot testing it after removing the frequent string comparisons. 

https://github.com/minetest/minetest/pull/7034#discussion_r167692535
The setting is evaluated in the StreamLogOutput constructor: https://github.com/minetest/minetest/pull/7034/files#diff-076b60230cbc7f307125f55094bd8a65R121
If the log streams are created before parsing the terminal arguments, the color mode is set to LOG_COLOR_AUTO (the static member's start value) in the constructor.
I assume this can be fixed by either testing the color mode every time a message appears or updating the color modes for existing output streams which are in m_outputs of the loggers.

**Comment by HybridDog** [comment]
Fixed: https://github.com/minetest/minetest/pull/7173

**Comment by paramat** [comment]
Blocker for 5.0.0?

**Comment by SmallJoker** [comment]
@paramat I don't think that's important enough to justify a blocker.

**Comment by nerzhul** [comment]
Already merged @paramat i don't see the debate here

**Pull request merged**